lib/rust/ensogl/component/grid-view/src/visible_area.rs

```diff
@@ -31,12 +32,24 @@ pub fn visible_rows(v: &Viewport, entry_size: Vector2, row_count: usize) -> Rang
 }
 
 /// Return range of visible columns.
-pub fn visible_columns(v: &Viewport, entry_size: Vector2, col_count: usize) -> Range<Col> {
-    let first_visible_unrestricted = (v.left / entry_size.x).floor() as isize;
-    let first_visible = first_visible_unrestricted.clamp(0, col_count as isize) as Col;
+pub fn visible_columns(
+    v: &Viewport,
+    entry_size: Vector2,
+    col_count: usize,
+    column_widths: &ColumnWidths,
+) -> Range<Col> {
+    let column_left = |idx: usize| idx as f32 * entry_size.x + column_widths.pos_offset(idx);
+    let column_right = |idx: usize| column_left(idx + 1);
+    let index = |(idx, _): (usize, _)| idx;
+
+    let visible = |(_, x): &(_, f32)| v.left < *x;
+    let mut right_borders = (0..col_count).map(|idx| (idx, column_right(idx)));
```

Comment:
I don't like iterating over all columns each time the viewport changes - it may change every frame. At least in the case where no column resizes occurs, it should be as fast as before. So let's count the "first_not_visible" as before and check the position offset; if it's big, then jump accordingly.

Reply:
I'm a bit confused. You refer to the `first_visible`-related code but mention `first_not_visible` in the comment. What do you mean by "as before"? `let first_visible = v.left / entry_size.x` is not possible anymore, because `entry_size.x` is variable now. There are cases when the first visible column is placed after some extended column, and we need to look through every previous column. 
I guess we can calculate the column's index that *should* be there and then check its `pos_offset`, and if it has any, then retreat back to the iterative solution. Is that your idea?

Reply:
> I'm a bit confused. You refer to the `first_visible`-related code but mention `first_not_visible` in the comment. What do you mean by "as before"? `let first_visible = v.left / entry_size.x` is not possible anymore, because `entry_size.x` is variable now. There are cases when the first visible column is placed after some extended column, and we need to look through every previous column. I guess we can calculate the column's index that _should_ be there and then check its `pos_offset`, and if it has any, then retreat back to the iterative solution. Is that your idea?

Yes, more or less. What I imagined is start for searching "first visible column" from the old "first_visible" value. And we check the current actual column position: if the column is still on the left viewport boundary, we're good. If not, we jump c.a. `(position_offset / entry_size.x) as isize` columns and check the new one.

This solution is not worse than iterative, but in case where only small amount of columns is resized it should look up for proper column very quickly

There is also better solution, but it requires changes in the SegmentPoint implementation: knowing the x you can traverse the tree from the root, going to the left children if the sum in it is less than x, and right child otherwise.